### PR TITLE
[Issue #507] Support Range based scan for Like Filters for No_Diction…

### PR DESCRIPTION
**Pull request by sujith71955**
This PR is basically filter optimization for conditional queries. Please find the optimization details mentioned below.

Range based scanning in case of no dictionary column is been supported for all types of conditional filters.
Logical conditional filters will support block pruning based on applied filter.

### CODE REVIEW

**Comment by sujith71955** [comment]
retest this please

**Comment by sujith71955** [comment]
retest this please

File: core/src/main/java/org/carbondata/query/carbon/executor/impl/AbstractQueryExecutor.java [hunk 1]
```diff
@@ -271,10 +271,12 @@ protected BlockExecutionInfo getBlockExecutionInfoForBlock(QueryModel queryModel
...
           .getFilterExecuterTree(queryModel.getFilterExpressionResolverTree(), blockKeyGenerator));
-      startIndexKey = queryModel.getFilterExpressionResolverTree()
-          .getstartKey(blockIndex.getSegmentProperties());
-      endIndexKey = queryModel.getFilterExpressionResolverTree()
-          .getEndKey(blockIndex.getSegmentProperties(), queryModel.getAbsoluteTableIdentifier());
+      List<IndexKey> listOfStartEndKeys = new ArrayList<IndexKey>(2);
+      FilterUtil.traverseResolverTreeAndGetStartAndEndKey(segmentProperties,
+          queryModel.getAbsoluteTableIdentifier(), queryModel.getFilterExpressionResolverTree(),
+          listOfStartEndKeys);
+      startIndexKey = listOfStartEndKeys.get(0);
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/carbon/executor/impl/AbstractQueryExecutor.java]: How about case if length is less than 2

File: core/src/main/java/org/carbondata/query/expression/conditional/ConditionalExpression.java [hunk 2]
```diff
@@ -31,6 +32,8 @@
 
   boolean isSingleDimension();
 
+  List<ExpressionResult> getExpressionResult();
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/expression/conditional/ConditionalExpression.java]: getLiterals

File: core/src/main/java/org/carbondata/query/expression/conditional/LessThanEqualToExpression.java [hunk 3]
```diff
@@ -26,7 +26,8 @@
 import org.carbondata.query.expression.ExpressionResult;
 import org.carbondata.query.expression.exception.FilterUnsupportedException;
 
-public class LessThanEqualToExpression extends BinaryConditionalExpression {
+public class LessThanEqualToExpression extends BinaryConditionalExpression
+    implements RangeExpression {
   private static final long serialVersionUID = 1L;
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/expression/conditional/LessThanEqualToExpression.java]: send correct serialization id

File: core/src/main/java/org/carbondata/query/expression/conditional/RangeExpression.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.carbondata.query.expression.conditional;
+
+public interface RangeExpression {
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/expression/conditional/RangeExpression.java]: add comment why this is required

File: core/src/main/java/org/carbondata/query/filter/executer/RangeFilterExecuterImpl.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+package org.carbondata.query.filter.executer;
+
+import java.util.BitSet;
+import java.util.List;
+
+import org.carbondata.core.util.ByteUtil;
+import org.carbondata.query.carbon.processor.BlocksChunkHolder;
+import org.carbondata.query.filter.resolver.resolverinfo.DimColumnResolvedFilterInfo;
+
+public class RangeFilterExecuterImpl implements FilterExecuter {
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/filter/executer/RangeFilterExecuterImpl.java]: this class is not required

File: core/src/main/java/org/carbondata/query/filter/executer/RowLevelFilterExecuterImpl.java [hunk 6]
```diff
@@ -84,12 +89,14 @@ public RowLevelFilterExecuterImpl(List<DimColumnResolvedFilterInfo> dimColEvalua
     }
 
     // CHECKSTYLE:OFF Approval No:Approval-V1R2C10_001
-    for (MeasureColumnResolvedFilterInfo msrColumnEvalutorInfo : msrColEvalutorInfoList) {
-      if (msrColumnEvalutorInfo.isMeasureExistsInCurrentSlice() && null == blockChunkHolder
-          .getMeasureDataChunk()[msrColumnEvalutorInfo.getColumnIndex()]) {
-        blockChunkHolder.getMeasureDataChunk()[msrColumnEvalutorInfo.getColumnIndex()] =
-            blockChunkHolder.getDataBlock().getMeasureChunk(blockChunkHolder.getFileReader(),
-                msrColumnEvalutorInfo.getColumnIndex());
+    if (null != msrColEvalutorInfoList) {
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/filter/executer/RowLevelFilterExecuterImpl.java]: this check it not required

File: core/src/main/java/org/carbondata/query/filter/executer/RowLevelRangeGrtThanFiterExecuterImpl.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      // so filter-max should be negative
+      int maxCompare =
+          ByteUtil.UnsafeComparer.INSTANCE.compareTo(filterValues[k], blockMaxValue[columnIndex]);
+      // and filter-min should be positive
+      int minCompare =
+          ByteUtil.UnsafeComparer.INSTANCE.compareTo(filterValues[k], blockMinValue[columnIndex]);
+
+      // if any filter value is in range than this block needs to be
+      // scanned
+      if (maxCompare < 0 && minCompare > 0) {
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/filter/executer/RowLevelRangeGrtThanFiterExecuterImpl.java]: > case only check (value < max) then scan
>= case only check (value <= max) then scan
< case only check (value > min) then scan
<= case only check (value >= min) then scan

File: core/src/main/java/org/carbondata/query/filters/FilterExpressionProcessor.java [hunk 8]
```diff
@@ -202,23 +212,55 @@ private void traverseAndResolveTree(FilterResolverIntf filterResolverTree,
...
         return new LogicalFilterResolverImpl(
-            createFilterResolverTree(currentExpression.getLeft(), tableIdentifier),
-            createFilterResolverTree(currentExpression.getRight(), tableIdentifier),
-            filterExpressionType);
+            createFilterResolverTree(currentExpression.getLeft(), tableIdentifier,
+                currentExpression),
+            createFilterResolverTree(currentExpression.getRight(), tableIdentifier,
+                currentExpression), filterExpressionType);
+
+      case RANGE:
```
Comment (gvramana) [inline core/src/main/java/org/carbondata/query/filters/FilterExpressionProcessor.java]: not required this

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonOperators.scala [hunk 9]
```diff
@@ -418,30 +479,41 @@ case class CarbonCubeScan(
...
+        val value = pattern.toString
+        val maxValueLimit = value.substring(0, value.size - 1) + value.charAt(value.size - 1) + 1
+        val r = new LessThanExpression(
+          transformExpression(left),
+          new CarbonLiteralExpression(maxValueLimit,
+            CarbonScalaUtil.convertSparkToCarbonDataType(dataType)
+          )
+        )
+        val logicalExp = new AndExpression(l, r)
+        logicalExp.setRangeExpression(true)
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/apache/spark/sql/CarbonOperators.scala]: this is not required and only this logic

File: integration/spark/src/test/scala/org/carbondata/spark/testsuite/detailquery/HighCardinalityDataTypesTestCase.scala [hunk 10]
```diff
@@ -74,6 +86,24 @@ class NO_DICTIONARY_COL_TestCase extends QueryTest with BeforeAndAfterAll {
 
   }
 
+  test("Detail Query with NO_DICTIONARY_COLUMN with Like range filter") {
+
+
+    checkAnswer(
+      sql("select empno from NO_DICTIONARY_CARBON_7 where empno like '12%'"),
```
Comment (gvramana) [inline integration/spark/src/test/scala/org/carbondata/spark/testsuite/detailquery/HighCardinalityDataTypesTestCase.scala]: No dictionary and filter, all operators add testcases

**Comment by sujith71955** [comment]
retest this please

**Pull request merged**